pkg/kubefed2/federate/federate.go

```diff
@@ -0,0 +1,281 @@
+// Kubernetes resource of the given kubernetes type.
+func NewCmdFederateResource(cmdOut io.Writer, config util.FedConfig) *cobra.Command {
+	opts := &federateResource{}
+
+	cmd := &cobra.Command{
+		Use:     "federate FEDERATED-TYPE-NAME RESOURCE-NAME",
+		Short:   "Federates a Kubernetes resource of the given type(identified by FederatedTypeConfig name)",
+		Long:    federate_long,
+		Example: federate_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := opts.Complete(args)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+
+			err = opts.Run(cmdOut, config)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+		},
+	}
+
+	flags := cmd.Flags()
+	opts.CommonBind(flags)
+	opts.Bind(flags)
+
+	return cmd
+}
+
+// Run is the implementation of the `federate resource` command.
+func (j *federateResource) Run(cmdOut io.Writer, config util.FedConfig) error {
+	hostConfig, err := config.HostConfig(j.HostClusterContext, j.Kubeconfig)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get host cluster config")
+	}
+
+	qualifiedTypeName := ctlutil.QualifiedName{
+		Namespace: j.FederationNamespace,
+		Name:      j.typeName,
+	}
+	qualifiedResourceName := ctlutil.QualifiedName{
+		Namespace: j.resourceNameSpace,
+		Name:      j.resourceName,
+	}
+
+	return FederateResource(hostConfig, qualifiedTypeName, qualifiedResourceName)
+}
+
+func FederateResource(hostConfig *rest.Config, qualifiedTypeName, qualifiedResourceName ctlutil.QualifiedName) error {
+	typeConfig, err := lookUpTypeDetails(hostConfig, qualifiedTypeName)
+	if err != nil {
+		return err
+	}
+
+	templateResource, err := getTargetResource(hostConfig, typeConfig, qualifiedResourceName)
+	if err != nil {
+		return err
+	}
+
+	return createFedResource(hostConfig, typeConfig, templateResource)
+}
+
+func lookUpTypeDetails(config *rest.Config, qualifiedTypeName ctlutil.QualifiedName) (*fedv1a1.FederatedTypeConfig, error) {
+	fedClient, err := util.FedClientset(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get federation clientset")
+	}
+
+	typeConfig, err := fedClient.CoreV1alpha1().FederatedTypeConfigs(qualifiedTypeName.Namespace).Get(qualifiedTypeName.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(err, "FederatedTypeConfig %q, does not exist. Type needs to be enabled first", qualifiedTypeName)
+		}
+		return nil, errors.Wrapf(err, "Error retrieving FederatedTypeConfig %q", qualifiedTypeName)
+	}
+
+	crdClient, err := apiextv1b1client.NewForConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating crd client")
+	}
+
+	_, err = crdClient.CustomResourceDefinitions().Get(typeconfig.GroupQualifiedName(typeConfig.GetFederatedType()), metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error retrieving Federated<type> CRD corresponding to FederatedTypeConfig %q. Federated<type> CRD should exist", qualifiedTypeName)
```

Comment:
The error message seems to assume that the only error would be due to a missing CRD.  I think that suggests a separate check for not found.

Also, consider using the name of the federated type rather than `Federated<type>`.

Reply:
Updated!